### PR TITLE
Removed namespaces as parameters and added launch file example in image_proc

### PR DESCRIPTION
**Pull request by caelinsutch**
For #515

### CODE REVIEW

File: image_proc/src/resize.cpp
Comment (SteveMacenski) [inline image_proc/src/resize.cpp]: I think you're missing a `/`?

File: image_proc/src/resize.cpp
Comment (SteveMacenski) [inline image_proc/src/resize.cpp]: You still have camera info hardcoded, these should be consistent.

@caelinsutch why not use remapping like we typically use in image pipeline? You can leave the `namespace/image` and then remap the `namespace/image` -> `namespace2/blah_image` rather than as a parameter. I think that's what you see in the existing launch file.

File: image_proc/src/resize.cpp
Comment (SteveMacenski) [inline image_proc/src/resize.cpp]: You shouldn't hardcode the `/` since its going to be gotten by a parameter. Parameters convention is not to include `/`.

File: image_proc/launch/image_proc.launch.py
Comment (SteveMacenski) [inline image_proc/launch/image_proc.launch.py]: Add extra line at end of files please

File: image_proc/launch/image_proc.launch.py
Comment (SteveMacenski) [inline image_proc/launch/image_proc.launch.py]: `nmea_serial_driver`?

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
It seems like you really want a remap rather than parameters. 

e.g. `image` is default and remap `image` -> `namespace/image_mono`. This would remove the namespace and the additional parameter you added. @klinton any reason not to do that?

File: image_proc/src/resize.cpp
Comment (caelinsutch) [reply image_proc/src/resize.cpp]: I was referencing the image_proc node's rectify.cpp that uses parameters for input and output, I'll change it to remaps.

File: image_proc/src/resize.cpp
Comment (SteveMacenski) [inline image_proc/src/resize.cpp]: Why remove this? This can be good to save on traffic if no one's actually listening.

File: image_proc/src/resize.cpp
Comment (SteveMacenski) [inline image_proc/src/resize.cpp]: Why change the default to `/camera`? Isn't `image` more standard?

File: image_proc/launch/image_proc.launch.py
Comment (SteveMacenski) [inline image_proc/launch/image_proc.launch.py:43]: I'm not trying to be difficult, but I don't understand. You remove the parameter, OK, but I don't see any remaps here to actually enable things to change those default values. 

It seems to be that what you want is no parameters (like in the original launch file here) and use remappings of the topics. It seems like the use of `camera_namespace` is strange. If you can remap the topic, you can include the namespace there, so there's no value in having a namespace parameter. 

How do your changes make this work where the original did not?

**Review by SteveMacenski — CHANGES_REQUESTED** [review]

File: image_proc/src/resize.cpp
Comment (caelinsutch) [reply image_proc/src/resize.cpp]: As mentioned in the other issue, there's a bug in the getNumSubscribers method that makes it always return 0. I fixed this here, your choice if you'd like me to put it back in

File: image_proc/src/resize.cpp
Comment (caelinsutch) [reply image_proc/src/resize.cpp]: This was a mistake on my end, I'll be fixing this

File: image_proc/launch/image_proc.launch.py
Comment (caelinsutch) [reply image_proc/launch/image_proc.launch.py:43]: I get what you're saying, I suppose I'm not sure of the value of this change, though I think an example with a remap would be valuable, which I'll be changing this too. What made my code work (on another project) was the remapping changes, which I didn't implement in this launch file. 

The `camera_namespace` keeps things a little more readable for me at least, and when used in remaps, prevents repetitive use of the same string for the namespace. I'll push an example tonight.

All in all, I don't mind the criticism, this is my first open source PR, so I'm still learning the ropes :), thanks for being accommodating!

File: image_proc/src/resize.cpp
Comment (SteveMacenski) [reply image_proc/src/resize.cpp]: Got it, can you please comment it out instead? that will be a placeholder to come back and address this issue.

File: image_proc/launch/image_proc.launch.py
Comment (SteveMacenski) [reply image_proc/launch/image_proc.launch.py:43]: @caelinsutch You can always make a launch file that has the namespace your pre-pend to the new remap target if you like (e.g. `{image, NAMESPACE + /remapped_name}`). But for the general case doing something like `{image, front_cam/image}` makes alot more sense to have remap abilities rather than independent namespace and image topic parameters. 

I think the right answer is to delete all the namespace parameters in image_proc components and then replace them with remaps like in the original launch file here. I think we added the namespaces here during a port before remap was available.

**Comment by SteveMacenski** [comment]
Also, make sure to take a look at the CI builder, it looks like you have linting problems. You can reproduce locally using `ament_flake8` command.

File: image_proc/launch/image_proc.launch.py
Comment (caelinsutch) [reply image_proc/launch/image_proc.launch.py:43]: So would you like me to do that in this PR, or make a new where I delete namespace params in the image_proc node

Comment (SteveMacenski) [reply image_proc/launch/image_proc.launch.py:43]: We should do it in this PR, I'm not sure this works as is without it. 

I think what should be done:
- Remove namespace parameter (then all pub/subs should be just "image" and "camera_info")
- Launch file has remaps for "image" and "camera_info" as needed for namespacing or changing to `_mono` `_color` for use in the pipeline

Agree?

Comment (caelinsutch) [reply image_proc/launch/image_proc.launch.py:43]: Sounds good, I'll likely write that up and write a PR for it tomorrow :)

Comment (caelinsutch) [reply image_proc/launch/image_proc.launch.py:43]: Updated with removed namespace for image_proc and added launch file remaps

**Comment by SteveMacenski** [comment]
I think you need to update the other image_proc items to use remap and not the camera namespace parameter. Otherwise, looks good.

**Comment by caelinsutch** [comment]
I'm not sure why the CI test is failing too, it says it has a problem w making but it makes fine on two of my machines :/. I'll update the other files in image_proc to use remap

**Comment by SteveMacenski** [comment]
https://github.com/ros-perception/image_pipeline/blob/ros2/.circleci/config.yml#L22

Try removing the parallel workers from the build and test.

**Review by SteveMacenski — APPROVED** [review]
Only thing I need to glance at is making sure all the namespaces are gone, but it looks like you got most/all of them.

**Comment by SteveMacenski** [comment]
Ugh, try `colcon build --parallel-workers 1`. Sorry, dropping the tag actually defaults to the number of cores (which was >2). My bad. 

Otherwise, looks good. Just needs to pass CI and verify that this works.

**Comment by SteveMacenski** [comment]
Cool, worked now. Failed from code lint errors

**Comment by caelinsutch** [comment]
@SteveMacenski How do I fix the uncrustify errors? I got rid of all the cpplint ones

**Comment by SteveMacenski** [comment]
Read the output of uncrustify and fix whatever issues they brought up. Not sure what else to say :smile:

**Comment by caelinsutch** [comment]
```
    --- src/crop_non_zero.cpp
    +++ src/crop_non_zero.cpp.uncrustify
    @@ -50,2 +50,2 @@
    -                                                std::bind(&CropNonZeroNode::imageCb,
    -                                                    this, std::placeholders::_1), "raw");
    +      std::bind(&CropNonZeroNode::imageCb,
    +      this, std::placeholders::_1), "raw");
 ```

Not sure what this means or how to fix it... The current code format looks nothing like this, I changed these lines to match the `-` lines (I think the old style?) but its been giving me the same errors. My current code doesn't look like the `+` lines at all, not sure where it's pulling these from

**Comment by JWhitleyWork** [comment]
@caelinsutch The `+` lines are the changes it wants you to make and the `-` lines are the originals that it wants you to remove. It's throwing this error because the `uncrustify` rules want continuation lines to only be indented one level deeper than the line above them (rather than indented all the way over to match the opening parentheses).

File: image_proc/launch/image_proc.launch.py
Comment (SteveMacenski) [inline image_proc/launch/image_proc.launch.py]: Can you verify for me that putting the `node_namespace` as `NAMESPACE` doesn't also namespace the topics? 

E.g. `('/image', NAMESPACE + '/image')` is that necessary with the `node_namespace=NAMESPACE`?

Comment (caelinsutch) [reply image_proc/launch/image_proc.launch.py]: I'll check that out tomorrow morning PST and get back to you then :)

Comment (SteveMacenski) [reply image_proc/launch/image_proc.launch.py]: If it turns out that its redundant, remove the node namespace and leave the remaps.

Comment (JWhitleyWork) [reply image_proc/launch/image_proc.launch.py]: If the topic names do not have a slash (which they shouldn't), then the `node_namespace` argument will namespace the topics.

Comment (SteveMacenski) [reply image_proc/launch/image_proc.launch.py]: It looks like many (if not all) have leading `/`s.

Comment (JWhitleyWork) [reply image_proc/launch/image_proc.launch.py]: That puts them in the root namespace which is bad juju. From http://wiki.ros.org/Names:

>Names that start with a "/" are global -- they are considered fully resolved. Global names should be avoided as much as possible as they limit code portability.

The "correct" way to handle topic names is to use un-prefixed names and let namespacing handle the full path resolution.

Comment (SteveMacenski) [reply image_proc/launch/image_proc.launch.py]: Yeah, agreed. I think for this demo the right answer is to remove the `/` and then do the remaps without the node_namespace set.

Comment (caelinsutch) [reply image_proc/launch/image_proc.launch.py]: I removed the namespace and changed all the topics to relative

File: image_proc/launch/image_proc.launch.py
Comment (SteveMacenski) [inline image_proc/launch/image_proc.launch.py:57]: `rectify_mono_node` but then you have the topic remap to `image_rect_color`. Looking at the other launch file, it looks like it has 2 rectify nodes: one for mono and the other for color.

File: image_proc/launch/image_proc.launch.py
Comment (SteveMacenski) [inline image_proc/launch/image_proc.launch.py]: You need to remove all leading `/` from both sides of the remap for all remaps.

File: image_proc/src/crop_non_zero.cpp
Comment (SteveMacenski) [inline image_proc/src/crop_non_zero.cpp]: This is actually what @JWhitleyWork meant, you're remapping these topics so these need to not have the `/` in front of them so that they could be namespaced if someone wanted. I see 10 instances of the `/` in the diff here and below. Just remove the leading `/` with no other alterations.

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
Playing wack-a-hole a little here. 
- Remove all `/` from all topics everywhere (see comments)
- Naming of the nodes mono/color. 

After that, good to go.

File: image_proc/src/crop_non_zero.cpp
Comment (caelinsutch) [reply image_proc/src/crop_non_zero.cpp]: Ah I see, thank's for the clarification!

File: image_proc/src/resize.cpp
Comment (SteveMacenski) [inline image_proc/src/resize.cpp]: Missed 2 here: image and resize

File: image_proc/launch/image_proc.launch.py
Comment (SteveMacenski) [inline image_proc/launch/image_proc.launch.py]:
```suggestion
                    ('image_rect', 'image_rect')
```

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
2 more `/` needing removed and a type just needs to accept my change

**Review by SteveMacenski — APPROVED** [review]
LGTM @JWhitleyWork any objections?

File: image_proc/launch/image_proc.launch.py
Comment (JWhitleyWork) [inline image_proc/launch/image_proc.launch.py:63]: These two remaps aren't actually remapping anything so they can be removed.

File: image_proc/launch/image_proc.launch.py
Comment (JWhitleyWork) [inline image_proc/launch/image_proc.launch.py]: This one can also be removed.

File: image_proc/launch/image_proc.launch.py
Comment (JWhitleyWork) [inline image_proc/launch/image_proc.launch.py]: You don't need the main function here. With launch files from Dashing forward, you just need to create the `generate_launch_description` function.

**Review by JWhitleyWork — CHANGES_REQUESTED** [review]
See comments on launch file. Otherwise looks good.

**Review by JWhitleyWork — APPROVED** [review]
Thanks for making those changes. Once you fix the test failures I'm good to merge.

File: image_proc/launch/image_proc.launch.py
Comment (SteveMacenski) [reply image_proc/launch/image_proc.launch.py:63]: @JWhitleyWork since this is a demo, I think its useful to show all 3 of them even if right now they don't do anything. Lets someone easily update this for their applications.

Comment (JWhitleyWork) [reply image_proc/launch/image_proc.launch.py:63]: Either way is fine with me. I'm not picky on this one - I just thought they were unnecessary. @caelinsutch, would you mind adding them back? Sorry about the back-and-forth.

**Comment by SteveMacenski** [comment]
@caelinsutch 
```
  <<< failure message
    -- run_test.py: invoking following command in '/root/project/src/image_pipeline/image_proc':
     - /opt/ros/dashing/bin/ament_flake8 --xunit-file /root/project/build/image_proc/test_results/image_proc/flake8.xunit.xml
    
    ./launch/image_proc.launch.py:33:1: F401 'sys' imported but unused
    import sys
    ^
    
    ./launch/image_proc.launch.py:35:1: F401 'launch.LaunchIntrospector' imported but unused
    from launch import LaunchDescription, LaunchIntrospector, LaunchService
    ^
    
    ./launch/image_proc.launch.py:35:1: F401 'launch.LaunchService' imported but unused
    from launch import LaunchDescription, LaunchIntrospector, LaunchService
    ^
    
    ./launch/image_proc.launch.py:36:1: F401 'launch_ros.get_default_launch_description' imported but unused
    from launch_ros import actions, get_default_launch_description
    ^
    
    4     F401 'sys' imported but unused
    
    1 files checked
    4 errors
    
    'F'-type errors: 4
    
    Checked files:
    
    * ./launch/image_proc.launch.py
    -- run_test.py: return code 1
    -- run_test.py: verify result file '/root/project/build/image_proc/test_results/image_proc/flake8.xunit.xml'
  >>>
build/image_proc/test_results/image_proc/flake8.xunit.xml: 4 tests, 0 errors, 4 failures, 0 skipped
- image_proc.flake8 F401 (./launch/image_proc.launch.py:33:1)
  <<< failure message
    'sys' imported but unused:
    import sys
  >>>
- image_proc.flake8 F401 (./launch/image_proc.launch.py:35:1)
  <<< failure message
    'launch.LaunchIntrospector' imported but unused:
    from launch import LaunchDescription, LaunchIntrospector, LaunchService
  >>>
- image_proc.flake8 F401 (./launch/image_proc.launch.py:35:1)
  <<< failure message
    'launch.LaunchService' imported but unused:
    from launch import LaunchDescription, LaunchIntrospector, LaunchService
  >>>
- image_proc.flake8 F401 (./launch/image_proc.launch.py:36:1)
  <<< failure message
    'launch_ros.get_default_launch_description' imported but unused:
    from launch_ros import actions, get_default_launch_description
  >>>
```

Run ament_flake8 locally to find these issues. You can also re-add the remaps to be complete as well, I'd appreciate that.

**Comment by JWhitleyWork** [comment]
> Run ament_flake8 locally to find these issues.

I recommend running `colcon test --packages-select image_proc && colcon test-result --verbose` to make sure it catches any and all test/linter failures.

**Comment by SteveMacenski** [comment]
Yes, but for just running the 1 test to quickly trouble shoot, you don't want to wait for all the tests to run to get feedback :-)

**Comment by SteveMacenski** [comment]
Alright, I'm happy with it. @JWhitleyWork merge if you're happy with it. 

Sorry for such a painful experience @caelinsutch.

**Pull request merged**
